book/chapter-01-unix-sockets/01_unix_sockets_intro.md

```diff
@@ -0,0 +1,292 @@
+### Datagrams
+
+Datagrams are also known as **connection-less** sockets: i.e. the other party never knows if the other has left. A datagram is a discrete message with boundaries - there is a beginning, body, and end to each one.
+
+A nice analogy I have stolen from [this Stackoverflow response](https://stackoverflow.com/a/4688899) makes it a lot easier to visualise:
+
+> A stream socket is like a phone call -- one side places the call, the other answers, you say hello to each other (SYN/ACK in TCP), and then you exchange information. Once you are done, you say goodbye (FIN/ACK in TCP). If one side doesn't hear a goodbye, they will usually call the other back since this is an unexpected event; usually the client will reconnect to the server. There is a guarantee that data will not arrive in a different order than you sent it, and there is a reasonable guarantee that data will not be damaged.
+
+> A datagram socket is like passing a note in class. Consider the case where you are not directly next to the person you are passing the note to; the note will travel from person to person. It may not reach its destination, and it may be modified by the time it gets there. If you pass two notes to the same person, they may arrive in an order you didn't intend, since the route the notes take through the classroom may not be the same, one person might not pass a note as fast as another, etc.
+
+> So you use a stream socket when having information in order and intact is important. File transfer protocols are a good example here. You don't want to download some file with its contents randomly shuffled around and damaged!
+
+**Note well**: An important distinction to make with Unix datagram packets (compared to _network sockets_) is that the order of the packets are guaranteed, and the order will remain the same - but the analogy still holds!
+
+## An example
+We can create a quick example of how this works using python; although any language (including `ba/sh`) would be more than suffice here, although python has some advantages we can get into later (specifically - looking at contexts)!
+
+```python
+# server.py
+import socket
+
+with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
+    sock.bind("/tmp/docode.sock")
+    sock.listen()
+    print(f"Listening on {sock.getsockname()}")
+    conn, address = sock.accept()
+    with conn:
+        while True:
+            data = conn.recv(1024)
+            print(f"Got connection with message: {data}")
+            if not data:
+                break
+            conn.sendall(data)
+```
+
+If we run `python server.py` we can see the address we are listening on - but the program will appear to hang: it is awaiting a connection before proceeding!
+
+We can prove the kernel is listenting on this socket by creating another terminal instance, and running
+
+```sh
+# the -l flag filters out those to ensure only sockets that are listenting are returned
+# the -p flag shows the associated process
+netstat -lp 
```

Comment:
for mac we have to use the following command:
```
sudo lsof -i -n -P
```
I think it's also important to warn the reader that this process can take a while as it's scanning through all the sockets, so they will have to wait for the result as opposed to thinking that their terminal has crashed